File: C++/프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

const int inf = 50000;
int min_[20001];
vector<int> edge_[20001];
int solution(int n, vector<vector<int>> edge) {
    for (int i = 1; i <= n; i++)
        min_[i] = inf;

    priority_queue<pair<int, int>> pq;
    for (auto it = edge.begin(); it != edge.end(); it++) {
        vector<int> &e = *it;
        edge_[e[0]].push_back(e[1]);
        edge_[e[1]].push_back(e[0]);
    }
    min_[1] = 0;
    pair<int, int> p(0, 1);
    pq.push(p);
    while (!pq.empty()) {
        p = pq.top();
        pq.pop();
        auto &it = edge_[p.second];
        for (int i = 0; i < it.size(); i++) {
            int dst = it[i];
            int dist = -p.first;
            if (dist + 1 < min_[dst]) {
                min_[dst] = dist + 1;
                pq.push(make_pair(-min_[dst], dst));
            }
        }
    }
    priority_queue<int> dis;
    for (int i = 1; i <= n; i++){
        dis.push(min_[i]);
    }
        int top = dis.top();
    int cnt = 0;
    while (!dis.empty() && dis.top() == top) {
        cnt++;
        dis.pop();
    };
    return cnt;
}

File: Java/프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public static List<LinkedList<Integer>> adjList;
    public static int[] dst;
    
    public int solution(int n, int[][] edge) {
        
        adjList = new ArrayList<>(n + 1);
        dst = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            adjList.add(new LinkedList<>());
            dst[i] = Integer.MAX_VALUE;
        }   
        
        for (int[] e : edge) {
            adjList.get(e[0]).add(e[1]);
            adjList.get(e[1]).add(e[0]);
        }
        
        dst[1] = 0;
        Queue<Dist> pq = new PriorityQueue<>();
        pq.add(new Dist(1, 0));
        
        while (!pq.isEmpty()) {
            Dist poll = pq.poll();
            for (int e : adjList.get(poll.d)) {
                if (poll.c + 1 < dst[e]) {
                    dst[e] = poll.c + 1;
                    pq.add(new Dist(e, dst[e]));
                }
            }
        }
        
        dst[0] = 0;
        int max = 0;
        for (int d : dst)
            max = Math.max(d, max);
        int answer = 0;
        for (int d : dst)
            answer += ((max == d) ? 1 : 0);
            
        return answer;
    }
    
    static class Dist implements Comparable {
        int d; int c;
        public Dist(int d_, int c_) { d = d_; c = c_;}
        
        @Override
        public int compareTo(Object o) {
            if (this.d > ((Dist)o).d)
                return 1;
            return -1;
        }
    }
}